pkg/node/node.go

```diff
@@ -244,6 +244,7 @@ func NewBee(addr string, swarmAddress swarm.Address, publicKey ecdsa.PublicKey,
 		return nil, fmt.Errorf("p2p service: %w", err)
 	}
 	b.p2pService = p2ps
+	defer p2ps.Ready()
```

Comment:
do we really need the `Ready` call if startup fails midway?

Reply:
I prefer since when we fail midway no shutdown will be called (this is wrong, as some things might already have happened by that point), and so some goroutines might already be waiting on that select case, and they should at least be released. Ideally as I previously mentioned we should also add a `Close` method on the libp2p library, and add a `quit` channel on the select case

Reply:
should we not perhaps cancel the `p2pCtx` in an error case? that would also trigger the select. also this context is passed into `libp2p.New` (which then passes it on to the real `libp2p.New` from libp2p), so this might also use this already to wait on something.

Reply:
If both ready is closed and context is canceled it is not deterministic which select case would happen in the handshake handler. That is not so problematic but still not the best thing to have. But, @ralph-pichler suggestion to cancel the context in error case is very good. I would add an argument to Ready function, for example bool and that the channel is of bool values, so that we can check if it is false, to return in the handler, and if it is true to unblock the handler. It is not so clear naming in that case, maybe to rename the method to something else, but Ready is ok. But what way the described problem with select case would not exist, as both context will be canceled and Ready would pass the false value, both resulting in the return in the handler.

Reply:
OK point taken :+1: